First iteration of Buzzhouse

PedroTadim opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->
The first iteration of Buzhouse is here, Buzzhouse is a new fuzzer aimed to stress test corner cases not handled by other existing CI fuzzers in Clickhouse.  Compile ClickHouse with `ENABLE_BUZZHOUSE=1` to enable it.

It starts by creating a few databases and tables, then sending random read and write queries while keeping the catalog information on them. So far, it covers many Clickhouse types, many SELECT query features, INSERTs, most mutations, many table engines, and a few other SQL features. A few query oracles have also been implemented to detect queries with wrong results. Buzzhouse will keep running until a wrong result is found or an assertion or logical error is raised by either the client or the server.

I have a long list of new testing features to implement, and they will be implemented over time. So far it has found about 60 bugs in Clickhouse.

Buzzhouse runs in CH's client must like AST fuzzer does. To start it, run:

```
clickhouse --client --buzz-house-config=<path to json configuration file>
```
The configuration file contains the following options:

- db_file_path - Path string with CH server directory where user files are stored for import and export statements
- log_path - Path string with the location of the file where the SQL query log will be located
- read_log - Boolean telling to read the file in `log_path` to reproduce a previous session (default false).
- seed - The seed value for the random generator. The default is 0, which means to use a random one from the system.
- max_depth - Max depth dor queries generated (default 3).
- max_width - Max width for queries generated (default 3).
- max_databases - Max number of databases created (default 4).
- max_functions - Max number of functions created (default 4).
- max_tables - Max number of tables created (default 10).
- max_views - Max number of views created, either materialized, refreshable, or neither (default 5).
- fuzz_floating_points - Create tables with Float and Double types. Beware that many false positives can be raised in the oracles for these types, so you may disable them (default true).
- test_with_fill - Generate `ORDER BY WITH FILL` queries. Notice some `WITH FILL` queries may take a long time (default true).
- use_dump_table_oracle - Dump and insert the contents of a table, then compare the contents. At the moment, it may give false positives for some data types such as Dynamic (default true).
- max_insert_rows - Max rows to be generated for insert queries (default 100).
- max_nested_rows - Max rows generated for nested types (default 10).

I may add other configuration options in the future.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [x] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [x] <!---ci_exclude_aarch64--> Exclude: All with aarch64
- [ ] <!---ci_exclude_release--> Exclude: All with release
- [ ] <!---ci_exclude_debug--> Exclude: All with debug
---
- [x] <!---ci_include_uzz--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, BuzzHouse, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [x] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #71085 from PedroTadim/ClickHouse:fuzz-test -->
*This is an automated comment for commit ea6d5bf35d916a7cc50f763dba23e343380b8347 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71085/ea6d5bf35d916a7cc50f763dba23e343380b8347/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c82cf25b3e5864bcc153cbe45adb8c6527e1ec6e/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

PedroTadim committed:
Geo types

PedroTadim committed:
Typo and follow same rules

PedroTadim committed:
Fix

PedroTadim committed:
Truncate database and fix

PedroTadim committed:
Fix and geo functions

PedroTadim committed:
More complex limit expressions

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
More settings

PedroTadim committed:
More partition functions and clang-format

PedroTadim committed:
Split table creation methods into their own translation unit

PedroTadim committed:
Clang format and settings

PedroTadim committed:
New settings and partition integers

PedroTadim committed:
Allow calling queries to the server from BuzzHouse itself

PedroTadim committed:
Allow limit and offset with ties for query oracles

PedroTadim committed:
Added missing empty cases

PedroTadim committed:
Adding partition alter and clang format

PedroTadim committed:
Add max and min date values

PedroTadim committed:
Fixes

PedroTadim committed:
Disable a few expression inside default values

PedroTadim committed:
Add more tries

PedroTadim committed:
Fixes

PedroTadim committed:
Alter partitions and many fixes

PedroTadim committed:
Drop, attach and detach parts

PedroTadim committed:
Fix

PedroTadim committed:
Use partitions in heavyweight deletes and updates

PedroTadim committed:
Partition identifiers may not be numbers

PedroTadim committed:
More partition alters

PedroTadim committed:
Parts are allowed for CHECK statements

PedroTadim committed:
Alter projection in partition

PedroTadim committed:
Fixes

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim commented on programs/client/CMakeLists.txt:
```diff
@@ -1,6 +1,22 @@
+PROTOBUF_GENERATE_CPP(GRAMMAR_SRC GRAMMAR_HDR buzz-house/ast/sql_grammar.proto)
```
It's used by other components from CH. You can search for it.

pamarcos replied:
Got it, I had to reconfigure with `ENABLE_PROTOBUF=1`.
If we're going to always require protobuf for the client, I suggest we drop a nice message to help us figure out what to enable and how.

We do this sort of thing in other places: https://github.com/ClickHouse/ClickHouse/blob/9f62518846ca945c11a653b46d2d3f3885ad7f40/contrib/aws-cmake/CMakeLists.txt#L9-L15

PedroTadim replied:
I guess I should introduce a new option to compile BuzzHouse or not

pamarcos replied:
That'd be cool if it can be enabled/disabled on demand, but at least it's always helpful to know what needs to be configured in CMake to make something work. Searching for internal dependencies when not stated explicitly is not as easy as a direct message.

pamarcos commented on programs/client/CMakeLists.txt:
```diff
@@ -1,6 +1,22 @@
+    buzz-house/third_party/uhugeint.cpp
+    ${GRAMMAR_SRC}
 )
+set_source_files_properties(${CLICKHOUSE_CLIENT_SOURCES} PROPERTIES COMPILE_FLAGS "-Wno-reserved-identifier -Wno-extra-semi-stmt -Wno-used-but-marked-unused -Wno-covered-switch-default")
```
Why did you have to disable all these warnings for buzz-house?

If it's for our own buzz-house code, I think we should stick to the same standards as the rest of the code.
If it's third-party code, we can use pragma statements to change diagnostics of just those files, or change the compiler flags only for those files.

PedroTadim replied:
Removed a few of them. These flags are needed because of Protobuf. Codegen fuzzer also sets them.

pamarcos commented on programs/client/Client.cpp:
```diff
@@ -1,31 +1,31 @@
+#include "Client.h"
```
I guess you're using some plugin to sort includes, change format the code and all that after saving, but it's making the review much more difficult, apart of not being consistent with the rest of the code.

Is there any way to leave the code as it was and only apply formatting to your actual changes?

PedroTadim replied:
I am applying clang-format according to ClickHouse's format, but I am not getting the supposed output.

pamarcos commented:
@PedroTadim it's a huge change in terms of LOC, so please help me to review it easier. What I'd like to know:

- Can we stick to the coding style of the rest of ClickHouse for our own code? Filenames don't use `snake_case` but `PascalCase`. The same for namespaces. Why is `buzzhouse` even in a different space outside of `DB`? Includes use `<>` instead of double quotes `""`. Check out the [coding style](https://clickhouse.com/docs/en/development/style) and mimic the code around.
- Which third-party libraries are you integrating, if any? What's their license? Shouldn't they be in the contrib dir instead of within buzzhouse?
- What files are auto-generated, if any? I'm guessing the AST ones, but it'd be good to have a README saying what's generated and how we can generate them again if we ever need it

:pray: Please keep your changes to me minimum without reformatting and sorting again includes unless it's necessary, because seems your editor is applying some formatting that it's not compliant with our coding style, even if it's using our `clang-format` conf file (which I guess may not stick to the coding style that well).

PedroTadim committed:
Use ClickHouse own json parser

PedroTadim committed:
Use OpenSSL's MD5 implementation

PedroTadim committed:
Remove third-party reference

PedroTadim commented:
> @PedroTadim it's a huge change in terms of LOC, so please help me to review it easier. What I'd like to know:
> 
>     * Can we stick to the coding style of the rest of ClickHouse for our own code? Filenames don't use `snake_case` but `PascalCase`. The same for namespaces. Why is `buzzhouse` even in a different space outside of `DB`? Includes use `<>` instead of double quotes `""`. Check out the [coding style](https://clickhouse.com/docs/en/development/style) and mimic the code around.
> 
>     * Which third-party libraries are you integrating, if any? What's their license? Shouldn't they be in the contrib dir instead of within buzzhouse?
> 
>     * What files are auto-generated, if any? I'm guessing the ones in `ast` and in `generator`, but it'd be good to have a README saying what's generated automatically and how we can generate them again if we ever need it.
> 
> 
> 🙏 Please keep your changes to the minimum without reformatting and sorting again includes unless it's necessary, because seems your editor is applying some formatting that it's not compliant with our coding style, even if it's using our `clang-format` conf file (which I guess may not stick to the coding style that well).

Removed remaining 3rd party code, now looking at clang-format

PedroTadim committed:
Following code conventions

PedroTadim committed:
Set CI job name

PedroTadim committed:
Fixing CI

PedroTadim committed:
New BFloat16 type

PedroTadim committed:
Disallow state combinatores while running query oracles

PedroTadim committed:
Missing settings

PedroTadim committed:
Use partition id, it allows non numeric partition names

PedroTadim committed:
Change previous commit, always use partition id for simplicity (only numbers)

PedroTadim committed:
Use camel case for function names

PedroTadim committed:
Fixing CI

PedroTadim committed:
Make external integration libraries optional

PedroTadim committed:
Remove commented out code

PedroTadim committed:
Remove nextRandomDouble, printing a FP value is too large

PedroTadim committed:
Avoid void casting

PedroTadim committed:
Fixing CI

PedroTadim committed:
Print messages when integration is missing

PedroTadim committed:
Fixing CI

PedroTadim committed:
Fixing CI

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Add CMake option ENABLE_BUZZHOUSE to build BuzzHouse or not

PedroTadim committed:
size_t vs uin32_t fixes

PedroTadim committed:
CI fix

PedroTadim committed:
Missing cast in C++ and for Dynamic type

pamarcos commented:
I think we need to find a better place for the `BuzzHouse` dir within `src`. `programs` seems not like the right place, because it only contains a the final client binary based on multiple core parts.

pamarcos commented on CMakeLists.txt:
```diff
@@ -88,6 +88,7 @@ string (TOUPPER ${CMAKE_BUILD_TYPE} CMAKE_BUILD_TYPE_UC)
 list(REVERSE CMAKE_FIND_LIBRARY_SUFFIXES)
 
 option (ENABLE_FUZZING "Fuzzy testing using libfuzzer" OFF)
+option (ENABLE_BUZZHOUSE "Enable BuzzHouse fuzzer on the client" OFF)
```
This is a great addition to allow building the client without `BuzzHouse`. However, to make sure our CI builds it, either we set it to ON by default, or we modify the build options to enable `BuzzHouse`

pamarcos replied:
Ok, I suggested to enable it here for now to make it easy, but my point was that I'd like to see the CI building the same thing we're both building on our dev machines to ensure it works.

So, feel free to modify the CI build instead to make sure this builds correctly.

PedroTadim committed:
Compilation fixes

PedroTadim committed:
Build buzzhouse in CI

PedroTadim committed:
Fix

pamarcos committed:
Fix some type warnings

pamarcos committed:
Build with ENABLE_BUZZHOUSE=1

PedroTadim committed:
Set a time for buzzhouse to run as a parameter

PedroTadim committed:
Run for three minutes by default

PedroTadim committed:
CI fixes and others

PedroTadim committed:
Fixes and missing clause

PedroTadim committed:
Fixes

PedroTadim committed:
CI Fix

PedroTadim committed:
Fixes

PedroTadim committed:
Make sure at least 3 tables/views/databases are available

pamarcos committed:
Add BuzzHouse build option for the CI

PedroTadim committed:
CI fixes

PedroTadim committed:
Fix

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Fixing CI

PedroTadim committed:
More fixes

PedroTadim committed:
Compatibility fix

PedroTadim committed:
Fixes

PedroTadim committed:
Used the wrong script

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
FIxing CI

PedroTadim committed:
Fixes and add table comments

pamarcos committed:
Ignore .pb.h headers in clang-tidy

pamarcos committed:
Merge branch 'master' into fuzz-test

pamarcos committed:
Fix to ignore the warnings for SQLGrammar.pb.h

pamarcos committed:
Merge branch 'master' into fuzz-test

PedroTadim committed:
Use global aggregate queries in query oracles

PedroTadim committed:
Fixing clang-tidy

PedroTadim committed:
CI fix

PedroTadim committed:
Clang-tidy fixes

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Fix

PedroTadim committed:
Missing settings

PedroTadim committed:
Allow settings in more places and fixes

PedroTadim committed:
More clang-tidy fixes

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Small fix

PedroTadim committed:
Fixing clang-tidy build

PedroTadim committed:
Fix

PedroTadim committed:
Compilation fix, make sed uint64_t

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

pamarcos committed:
Fix fuzzer arguments so they're split

Before, all arguments were passed as a single argument

pamarcos committed:
Fix linting

PedroTadim committed:
Adding new peer table oracle and fixes

PedroTadim committed:
Fixing CI

pamarcos committed:
Revert unnecessary changes

pamarcos committed:
Store fuzzer_out.sql as an artifact

pamarcos committed:
Reduce the fuzzing timeout to 1m to test the CI

pamarcos committed:
Fix linting

pamarcos committed:
Fix fuzzers not running

Ok, this is a very nasty thing. For some reason, we're hitting
some limits when passing that many arguments when running the
process in the background.

The same command running in the foreground does work perfectly,
but we need to do some workaround to make it work in the background.

pamarcos committed:
Revert "Reduce the fuzzing timeout to 1m to test the CI"

This reverts commit 380abbc67637510410815a66b363cefedd8bbc5c.

PedroTadim committed:
Implemented peer table oracle, now test

PedroTadim committed:
CI fixes

PedroTadim committed:
Make loop less heavy

PedroTadim committed:
Compilation fixes

PedroTadim committed:
Fix

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Fixes

PedroTadim committed:
Multiple fixes

PedroTadim committed:
Fix

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Adding new testing features

PedroTadim committed:
Generate system statements

PedroTadim committed:
Disable code

PedroTadim committed:
Fetch existing storage policies

PedroTadim committed:
Fix

PedroTadim committed:
Follow the convention

PedroTadim committed:
This is still needed

PedroTadim committed:
SharedSet and SharedJoin tables do exist

PedroTadim committed:
Use table settings in views

PedroTadim committed:
Fix

PedroTadim committed:
Update probabilities

PedroTadim committed:
Fix

PedroTadim committed:
Correct way to set timezones

PedroTadim committed:
Move partitions between disks

PedroTadim committed:
Create more databases

PedroTadim committed:
Revert change

PedroTadim committed:
Fixes and new feature

PedroTadim committed:
Revert change

PedroTadim committed:
New staleness feature

PedroTadim committed:
Fixes and new settings

PedroTadim committed:
Fixes

pamarcos committed:
Remove duplicated artifacts

The Python CI script already takes them into
account to generate the artifacts list.

pamarcos committed:
Fix fuzzer args

pamarcos committed:
Fix fuzzer args for many arguments

pamarcos committed:
Fix style

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

pamarcos committed:
Fix for real the arg issue

Turns out that timeout takes a while to spawn the process
when the number of arguments is huge. I wonder how this
ever worked before. Maybe it was flaky?

pamarcos committed:
Fix

pamarcos committed:
Fixes and new settings

pamarcos committed:
Fix error when compiling BuzzHouse without SQLite

pamarcos committed:
Merge branch 'master' into fuzz-test

pamarcos committed:
Add all fuzzers when including only them

ci_include_ matches whatever is after that, as opposed to
ci_exlucde_ where you can exclude several words separate by `|`.

pamarcos committed:
Remove unnecessary report

This is redundant given the generic report

pamarcos committed:
Fix fuzzer artifacts

pamarcos committed:
Merge branch 'master' into fuzz-test

pamarcos committed:
Fix run-fuzzer after removal of unneeded report.html

pamarcos committed:
Fix

pamarcos committed:
More settings

pamarcos committed:
Bring back the 30min of fuzzing

pamarcos committed:
Added mask value for int64 and fixes

pamarcos committed:
Give 30s of grace period for the client to finish

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Fixes

PedroTadim committed:
Adding TTL expressions to tables and columns

PedroTadim committed:
Fixes

PedroTadim committed:
Fixes

PedroTadim committed:
Added 24.12 features

PedroTadim committed:
clang-tidy fixes

PedroTadim committed:
More clang-tidy fixes

PedroTadim committed:
Fixes

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
My bad

PedroTadim committed:
Clang-tidy fixes

PedroTadim committed:
Several fixes

PedroTadim committed:
Fix

PedroTadim committed:
MIssing column and fixes

PedroTadim committed:
More CI fixes

PedroTadim committed:
Fixes

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Improving nested type generation

PedroTadim committed:
Fix

PedroTadim committed:
Flat json subcols and fixes

PedroTadim committed:
MIssing option and CI fix

PedroTadim committed:
Crucial fix

PedroTadim committed:
Fixes

PedroTadim committed:
Fixes and add external sources functions into regular queries

PedroTadim committed:
Fix

PedroTadim committed:
Small fix for views and CI

PedroTadim committed:
More CI fixes

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Trying to fix S3 function call

PedroTadim committed:
Fix for many nested cases

PedroTadim committed:
Forgot to update

pamarcos commented on programs/client/BuzzHouse/Generator/ExternalIntegrations.h:
```diff
@@ -0,0 +1,953 @@
+#include "FuzzConfig.h"
+#include "SQLCatalog.h"
+
+#if defined USE_MYSQL && USE_MYSQL
+#    if __has_include(<mysql.h>)
+#        include <mysql.h>
+#    else
+#        include <mysql/mysql.h>
+#    endif
+#endif
+
+#if defined USE_MONGODB && USE_MONGODB
+#    include <bsoncxx/builder/stream/array.hpp>
+#    include <bsoncxx/builder/stream/document.hpp>
+#    include <bsoncxx/json.hpp>
+#    include <bsoncxx/types.hpp>
+#    include <mongocxx/client.hpp>
+#    include <mongocxx/collection.hpp>
+#    include <mongocxx/database.hpp>
+#endif
+
+#if defined USE_LIBPQXX && USE_LIBPQXX
+#    include <pqxx/pqxx>
+#endif
+
+#if defined USE_SQLITE && USE_SQLITE
+#    include <sqlite3.h>
+#endif
```
I think there's no need to check whether things are defined or not, because they should always be defined in the config.h

```suggestion
#if USE_MYSQL
#    if __has_include(<mysql.h>)
#        include <mysql.h>
#    else
#        include <mysql/mysql.h>
#    endif
#endif

#if USE_MONGODB
#    include <bsoncxx/builder/stream/array.hpp>
#    include <bsoncxx/builder/stream/document.hpp>
#    include <bsoncxx/json.hpp>
#    include <bsoncxx/types.hpp>
#    include <mongocxx/client.hpp>
#    include <mongocxx/collection.hpp>
#    include <mongocxx/database.hpp>
#endif

#if USE_LIBPQXX
#    include <pqxx/pqxx>
#endif

#if USE_SQLITE
#    include <sqlite3.h>
#endif
```

PedroTadim replied:
Sometimes, they may be defined as 0, so double-checking is good.

pamarcos replied:
>Sometimes, they may be defined as 0, so double-checking is good.

Of course you need to check whether they're 0 or not, but you don't need to check if they're defined. They're always defined, and their value is either 1 or 0.

PedroTadim replied:
From here: https://stackoverflow.com/questions/17160755/is-there-a-way-to-both-check-a-macro-is-defined-and-it-equals-a-certain-value-at there's no straight way to do it

pamarcos commented on programs/client/BuzzHouse/Generator/ExternalIntegrations.h:
```diff
@@ -0,0 +1,953 @@
+namespace BuzzHouse
+{
+
+using IntegrationCall = enum IntegrationCall { IntMySQL = 1, IntPostgreSQL = 2, IntSQLite = 3, IntRedis = 4, IntMongoDB = 5, IntMinIO = 6 };
```
We should be using an `enum class` instead. Among others, that'd allow us to remove the `Int` prefix of all names, since they won't clash.

```suggestion
using IntegrationCall = enum class IntegrationCall { MySQL = 1, PostgreSQL = 2, SQLite = 3, Redis = 4, MongoDB = 5, MinIO = 6 };
```

Do not accept the suggestion as it is, because you'll need to manually refactor all references in other files.

PedroTadim replied:
MySQL, PostgeSQL and SQLite names already exist in BuzzHouse namespace, they need something different here.

pamarcos replied:
AFAIK, if you use an `enum class` you can call them internally however you want

pamarcos commented on programs/client/BuzzHouse/Generator/ExternalIntegrations.h:
```diff
@@ -0,0 +1,953 @@
+
+#include "config.h"
+
+#include "FuzzConfig.h"
+#include "SQLCatalog.h"
```
Check these includes and the rest. We use `<>` instead of double quotes `"`

pamarcos commented on programs/client/BuzzHouse/Generator/ExternalIntegrations.h:
```diff
@@ -0,0 +1,953 @@
+
+    virtual std::string getTableName(uint32_t) { return std::string(); }
+
+    virtual void getTypeString(RandomGenerator &, SQLType *, std::string &) const { }
```
I'd expect from any `get` method to return something. Seems you're returning the result in the 3rd parameter? I've seen that throughout the code, but I don't understand why having an out parameter is any better than returning the value directly for your use cases.

PedroTadim replied:
avoid allocating strings, that's it

PedroTadim committed:
Check if tables are available

PedroTadim committed:
Use a less confusing method name

PedroTadim committed:
DIsable dump oracle in CI

PedroTadim committed:
Set a limit to the numbr of rows in inserts

PedroTadim committed:
Use same size

PedroTadim committed:
Remove redundant setting

PedroTadim committed:
CI fix

PedroTadim committed:
Use better names

PedroTadim committed:
Fixing includes

PedroTadim committed:
More inserts

PedroTadim committed:
Use review suggestion for macro checks

Co-authored-by: Pablo Marcos <pablo.marcos.oltra@gmail.com>

PedroTadim committed:
Add space when needed

pamarcos committed:
Move BuzzHouse and Fuzzing mode to config.h

PedroTadim committed:
FIxing potential leaks

pamarcos committed:
Use enum class

pamarcos committed:
fixup! Use enum class

PedroTadim committed:
Do a proper copy

pamarcos committed:
Increase grace period and allow BuzzHouse to fail

Seems with gdb attached the process is taking way longer
than anticipated to handle SIGTERM.

Test fuzzing only for 2 minutes to see what the CI reports.

pamarcos committed:
Dummy commit to trigger the CI

pamarcos committed:
Another CI trigger

pamarcos commented:
@PedroTadim I finally got the fuzzers working at https://s3.amazonaws.com/clickhouse-test-reports/71085/662ccb9ad91e6d9be6bbb01dd116262c1de5597b/ci_running.html :raised_hands: 

I'll increase from 2min to 30min the time for running fuzzers, merge master and enable back all checks, alright?

pamarcos committed:
Run fuzzers for 30min

pamarcos committed:
Merge branch 'master' into fuzz-test

PedroTadim committed:
Adding system tables

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Added cloud tables

PedroTadim committed:
Fix for query oracle

PedroTadim committed:
Don't leak

PedroTadim committed:
Don't output into terminal

PedroTadim committed:
Fix

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Remove ambiguous setting

pamarcos committed:
Ensure the AST Fuzzer job never fails running BuzzHouse

Let's ignore errors found during BuzzHouse run for now.

PedroTadim committed:
Ignore stack trace table

pamarcos commented:
> I think we need to find a better place for the `BuzzHouse` dir within `src`. `programs` seems not like the right place, because it only contains a the final client binary based on multiple core parts.

I think we should move it from `programs/client/BuzzHouse` to `src/Client/Buzzhouse`

pamarcos commented on programs/client/BuzzHouse/Utils/Hugeint.h:
```diff
@@ -0,0 +1,78 @@
+// Forward declaration to allow conversion between hugeint and uhugeint
+struct uhugeint_t;
+
+struct hugeint_t
```
This class should be named `HugeInt`, in a `HugeInt.h` file (notice the uppercase i)

pamarcos commented on programs/client/BuzzHouse/Utils/UHugeint.h:
```diff
@@ -0,0 +1,78 @@
+// Forward declaration to allow conversion between hugeint and uhugeint
+struct hugeint_t;
+
+struct uhugeint_t
```
This class should be named `UHugeInt`, in a `UHugeInt.h` file (notice the uppercase i)

pamarcos commented on programs/client/BuzzHouse/Utils/UHugeint.h:
```diff
@@ -0,0 +1,78 @@
+    bool operator!() const;
+};
+
+template <>
+struct NumericLimits<uhugeint_t>
+{
+    static constexpr uhugeint_t Minimum() { return {0, 0}; }
+    static constexpr uhugeint_t Maximum() { return {std::numeric_limits<uint64_t>::max(), std::numeric_limits<uint64_t>::max()}; }
+};
```
This seems completely unused

pamarcos commented on programs/client/BuzzHouse/Utils/Hugeint.h:
```diff
@@ -0,0 +1,78 @@
+template <>
+struct NumericLimits<hugeint_t>
+{
+    static constexpr hugeint_t Minimum() { return {std::numeric_limits<int64_t>::lowest(), 0}; }
+    static constexpr hugeint_t Maximum() { return {std::numeric_limits<int64_t>::max(), std::numeric_limits<uint64_t>::max()}; }
```
Function members should start in lowercase. Thus, should be `minimum` and `maximum`

yakov-olkhovskiy committed:
adjustments for having port in forwarded_for field

robot-clickhouse committed:
Automatic style fix

PedroTadim committed:
Address review comments

pamarcos commented on programs/client/BuzzHouse/Utils/Hugeint.cpp:
```diff
@@ -0,0 +1,601 @@
+#include <Hugeint.h>
+#include <UHugeint.h>
```
These files should be called `HugeInt.h/cpp` and `UHugeInt.h/cpp`. Notice the uppercase `I`

PedroTadim committed:
More review comments

PedroTadim committed:
Ignore more settings

PedroTadim committed:
Ignore another setting

PedroTadim committed:
Moved BuzzHouse src into src/Client

PedroTadim commented:
> > I think we need to find a better place for the `BuzzHouse` dir within `src`. `programs` seems not like the right place, because it only contains a the final client binary based on multiple core parts.
> 
> I think we should move it from `programs/client/BuzzHouse` to `src/Client/Buzzhouse`

Just moved it.

PedroTadim committed:
Don't use CollapsingMergeTree tables for dump oracle

PedroTadim committed:
Use logging

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Attempting to fix clang-tidy build

PedroTadim committed:
Still fixing CI build

PedroTadim committed:
Fixing possible leaks

PedroTadim committed:
Trying to fix build

PedroTadim committed:
Add parallel replicas settings

PedroTadim committed:
Set wise limits

PedroTadim committed:
Merge remote-tracking branch 'upstream/master' into fuzz-test

PedroTadim committed:
Adding Merge Engine

PedroTadim committed:
New settings

PedroTadim committed:
Fix CI and missing setting

PedroTadim committed:
Real fix
